harness/determined/pytorch/experimental/_torch_batch_process.py

```diff
@@ -0,0 +1,433 @@
+        self._use_default_storage = True
+        return self._core_context.checkpoint._storage_manager.store_path(self._storage_path)
+
+
+def get_default_device(core_context: core.Context) -> torch.device:
+    local_rank = core_context.distributed.local_rank
+    local_num_gpu = torch.cuda.device_count()
+    if local_num_gpu == 0:
+        return torch.device("cpu")
+    else:
+        # Assuming there would not be more process than CUDA devices
+        return torch.device("cuda", local_rank)
+
+
+def _initialize_distributed_backend() -> Optional[core.DistributedContext]:
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+
+    info = det.get_cluster_info()
+    if info and (len(info.container_addrs) > 1 or len(info.slot_ids) > 1):
+        raise ValueError(
+            "In multi-slot managed cluster training, you must wrap your training script with a "
+            "distributed launch layer such as determined.launch.torch_distributed"
+        )
+    return None
+
+
+def _initialize_default_inference_context() -> core.Context:
+    distributed_context = _initialize_distributed_backend()
+    # Use WorkerAskChief mode to ensure synchronize correctly across worker
+    # Using WorkerAskMaster mode could lead to some workers exiting when others
+    # are waiting for synchronization.
+    return det.core.init(
+        distributed=distributed_context, preempt_mode=core.PreemptMode.WorkersAskChief
+    )
+
+
+class TorchBatchProcessor(metaclass=abc.ABCMeta):
+    """
+    User can initialize necessary resources in the init function, such as
+    - model for prediction
+    - storage client (e.g. s3 client)
+    """
+
+    @abc.abstractmethod
+    def __init__(self, context: TorchBatchProcessorContext) -> None:
+        pass
+
+    @abc.abstractmethod
+    def process_batch(self, batch: Any, batch_idx: int) -> None:
+        pass
+
+    def on_checkpoint_start(self) -> None:  # noqa: B027
+        """
+        Overwrite this function to run certain logic before checkpointing
+        This function will be called right before each checkpoint.
+        """
+        pass
+
+    def on_finish(self) -> None:  # noqa: B027
+        """
+        This function will be called right before exiting after completing iteration
+        over dataset
+        """
+        pass
+
+
+def _load_state(checkpoint_directory: pathlib.Path) -> Any:
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath("metadata.json").open("r") as f:
+        metadata = json.load(f)
+        return metadata
+
+
+def _synchronize_and_checkpoint(
+    core_context: core.Context, steps_completed: int, rank: int, default_output_uuid: str
```

Comment:
why pass core_context and also rank, since rank can be derived from `core_context.distributed.rank`?